core/src/main/scala/kafka/server/ConfigAdminManager.scala

```diff
@@ -0,0 +1,518 @@
+                throw new ClusterAuthorizationException(Errors.CLUSTER_AUTHORIZATION_FAILED.message())
+              }
+              validateResourceNameIsCurrentNodeId(resource.resourceName())
+              validateLogLevelConfigs(resource.configs())
+              if (!request.validateOnly()) {
+                alterLogLevelConfigs(resource.configs())
+              }
+              results.put(resource, ApiError.NONE)
+            case BROKER =>
+              // The resource name must be either blank (if setting a cluster config) or
+              // the ID of this specific broker.
+              if (!configResource.name().isEmpty) {
+                validateResourceNameIsCurrentNodeId(resource.resourceName())
+              }
+              validateBrokerConfigChange(resource, configResource)
+            case TOPIC =>
+            // Nothing to do.
+            case _ =>
+              throw new InvalidRequestException(s"Unknown resource type ${resource.resourceType().toInt}")
+          }
+        } catch {
+          case t: Throwable => {
+            val err = ApiError.fromThrowable(t)
+            info(s"Error preprocessing incrementalAlterConfigs request on ${configResource}", t)
+            results.put(resource, err)
+          }
+        }
+      }
+    })
+    results
+  }
+
+  def validateBrokerConfigChange(
+    resource: IAlterConfigsResource,
+    configResource: ConfigResource
+  ): Unit = {
+    val perBrokerConfig = !configResource.name().isEmpty
+    val persistentProps = configRepository.config(configResource)
+    val configProps = conf.dynamicConfig.fromPersistentProps(persistentProps, perBrokerConfig)
+    val alterConfigOps = resource.configs().asScala.map {
+      case config =>
+        val opType = AlterConfigOp.OpType.forId(config.configOperation())
+        if (opType == null) {
+          throw new InvalidRequestException(s"Unknown operations type ${config.configOperation}")
+        }
+        new AlterConfigOp(new ConfigEntry(config.name(), config.value()), opType)
+    }.toSeq
+    prepareIncrementalConfigs(alterConfigOps, configProps, KafkaConfig.configKeys)
+    try {
+      validateBrokerConfigChange(configProps, configResource)
+    } catch {
+      case t: Throwable => error(s"validation of configProps ${configProps} for ${configResource} failed with exception", t)
+        throw t
+    }
+  }
+
+  def validateBrokerConfigChange(
+    props: Properties,
+    configResource: ConfigResource
+  ): Unit = {
+    try {
+      conf.dynamicConfig.validate(props, !configResource.name().isEmpty)
+    } catch {
+      case e: ApiException => throw e
+      // TODO: InvalidRequestException is not really the right exception here if the
+      // configuration fails validation. The configuration is still well-formed, but just
+      // can't be applied. It should probably throw InvalidConfigurationException. However,
+      // we should probably only change this in a KIP since it has compatibility implications.
```

Comment:
Hmm. Do you think that we should file an issue for this and remove this TODO? I am worried that we are going to forget about this TODO.

Reply:
I'd prefer to keep it here since it explains why we are through an exception that doesn't seem appropriate. I will also create a JIRA for this.

https://issues.apache.org/jira/browse/KAFKA-13609